drivers/virtio/net/virtio_net.c

```diff
@@ -525,30 +536,41 @@ static int virtio_netdev_rxq_enqueue(struct virtio_net_device *vndev,
 		return -ENOSPC;
 	}
 
-	/**
-	 * Saving the buffer information before reserving the header space.
-	 */
-	buf_start = netbuf->data;
-	buf_len = netbuf->len;
-
-	/**
-	 * Retrieve the buffer header length.
-	 */
-	rc = uk_netbuf_header(netbuf, VTNET_HDR_SIZE_PADDED(vndev));
-	if (unlikely(rc != 1)) {
-		uk_pr_err("Failed to allocate space to prepend virtio header\n");
-		return -EINVAL;
-	}
-	rxhdr = netbuf->data;
-
+	/* Reset the scatter gather list */
 	sg = &rxq->sg;
 	uk_sglist_reset(sg);
 
-	/* Appending the header buffer to the sglist */
-	uk_sglist_append(sg, rxhdr, virtio_net_hdr_size(vndev));
+	if (vndev->buf_descr_count == VIRTIO_NET_BUF_DESCR_COUNT_INLINE) {
+		rc = uk_netbuf_header(netbuf, virtio_net_hdr_size(vndev));
+		if (unlikely(rc != 1)) {
+			uk_pr_err("Failed to allocate space to prepend virtio header\n");
+			return -EINVAL;
+		}
+		uk_sglist_append(sg, netbuf->data, netbuf->len);
```

Comment:
A lot of sglist appends whose return value is not checked around this driver... :thinking:

Reply:
I think I was adapting the new code a bit too much to the existing one. But yeah that is indeed a bit dangerous. I think it's because the driver predates paging and without paging you never have to worry about running out of sglist elements.

Reply:
Yeah maybe not worth bothering with this detail now. I am fine whatever you choose, I just hope we do not end up at some point debugging a difficult bug that we cannot pinpoint because we did not have an error check here. The way this whole driver looks is burning my eyes.